packages/inventory/src/services/reservation-item.ts

```diff
@@ -0,0 +1,269 @@
+      throw new MedusaError(
+        MedusaError.Types.NOT_FOUND,
+        `"reservationItemId" must be defined`
+      )
+    }
+
+    const manager = this.getManager()
+    const reservationItemRepository = manager.getRepository(ReservationItem)
+
+    const query = buildQuery({ id: reservationItemId }, config)
+    const [reservationItem] = await reservationItemRepository.find(query)
+
+    if (!reservationItem) {
+      throw new MedusaError(
+        MedusaError.Types.NOT_FOUND,
+        `ReservationItem with id ${reservationItemId} was not found`
+      )
+    }
+
+    return reservationItem
+  }
+
+  /**
+   * Create a new reservation item.
+   * @param data - The reservation item data.
+   * @return The created reservation item.
+   */
+  async create(data: CreateReservationItemInput): Promise<ReservationItem> {
+    const result = await this.atomicPhase_(async (manager) => {
+      const itemRepository = manager.getRepository(ReservationItem)
+
+      const inventoryItem = itemRepository.create({
+        inventory_item_id: data.inventory_item_id,
+        line_item_id: data.line_item_id,
+        location_id: data.location_id,
+        quantity: data.quantity,
+        metadata: data.metadata,
+      })
+
+      const [newInventoryItem] = await Promise.all([
+        itemRepository.save(inventoryItem),
+        this.inventoryLevelService_
+          .withTransaction(manager)
+          .adjustReservedQuantity(
+            data.inventory_item_id,
+            data.location_id,
+            data.quantity
+          ),
+      ])
+
+      return newInventoryItem
+    })
+
+    await this.eventBusService_.emit(ReservationItemService.Events.CREATED, {
+      id: result.id,
+    })
+
+    return result
+  }
+
+  /**
+   * Update a reservation item.
+   * @param reservationItemId - The reservation item's id.
+   * @param data - The reservation item data to update.
+   * @return The updated reservation item.
+   */
+  async update(
+    reservationItemId: string,
+    data: Omit<
+      DeepPartial<ReservationItem>,
+      "id" | "created_at" | "metadata" | "deleted_at"
+    >
+  ): Promise<ReservationItem> {
+    const item = await this.atomicPhase_(async (manager) => {
+      const itemRepository = manager.getRepository(ReservationItem)
+
+      const item = await this.retrieve(reservationItemId)
+
+      const shouldUpdateQuantity =
+        isDefined(data.quantity) && data.quantity !== item.quantity
+
+      const ops: Promise<unknown>[] = []
+      if (shouldUpdateQuantity) {
+        const quantityDiff = data.quantity! - item.quantity
+        ops.push(
+          this.inventoryLevelService_
+            .withTransaction(manager)
+            .adjustReservedQuantity(
+              item.inventory_item_id,
+              item.location_id,
+              quantityDiff
+            )
+        )
+      }
+
+      itemRepository.merge(item, data)
+      ops.push(itemRepository.save(item))
+
+      await Promise.all(ops)
+
+      return item
```

Comment:
question: wouldn't this return the item returned by `retrieve` i.e. before any updates occurred? I think we should return the result from `itemRepository.save` instead.

If the item is mutated by `merge` and `save` I think we should leave a comment, or take the performance hit in favor of readability and less side-effect heavy programming.

Reply:
good catch. We have to return the merged item + data